lib/storage/src/content_manager/consensus/consensus_wal.rs

```diff
@@ -1,203 +1,288 @@
-use std::fs::create_dir_all;
 use std::path::Path;
+use std::{cmp, fs};
 
-use itertools::Itertools;
 use prost_for_raft::Message;
 use protobuf::Message as _;
 use raft::eraftpb::Entry as RaftEntry;
 use wal::Wal;
 
 use crate::content_manager::consensus_manager;
-use crate::{ConsensusOperations, StorageError};
+use crate::content_manager::consensus_ops::ConsensusOperations;
+use crate::StorageError;
 
 const COLLECTIONS_META_WAL_DIR: &str = "collections_meta_wal";
 
-pub struct ConsensusOpWal(Wal);
+#[derive(Debug)]
+pub struct ConsensusOpWal {
+    wal: Wal,
+}
 
 impl ConsensusOpWal {
     pub fn new(storage_path: &str) -> Self {
         let collections_meta_wal_path = Path::new(storage_path).join(COLLECTIONS_META_WAL_DIR);
-        create_dir_all(&collections_meta_wal_path)
-            .expect("Can't create Collections meta Wal directory");
-        let wal = Wal::open(collections_meta_wal_path).expect("Can't open Collections meta Wal");
-        ConsensusOpWal(wal)
+
+        fs::create_dir_all(&collections_meta_wal_path)
+            .expect("Can't create consensus WAL directory");
+
+        let wal = Wal::open(collections_meta_wal_path).expect("Can't open consensus WAL");
+        Self { wal }
     }
 
     pub fn clear(&mut self) -> Result<(), StorageError> {
-        Ok(self.0.clear()?)
+        self.wal.clear()?;
+        Ok(())
     }
 
-    pub fn entry(&self, id: u64) -> raft::Result<RaftEntry> {
+    pub fn entry(&self, raft_index: u64) -> raft::Result<RaftEntry> {
         // Raft entries are expected to have index starting from 1
-        if id < 1 {
+        if raft_index < 1 {
             return Err(raft::Error::Store(raft::StorageError::Unavailable));
         }
-        let first_entry = self
-            .first_entry()
-            .map_err(consensus_manager::raft_error_other)?
-            .ok_or(raft::Error::Store(raft::StorageError::Unavailable))?;
-        if id < first_entry.index {
-            return Err(raft::Error::Store(raft::StorageError::Compacted));
-        }
-        // Due to snapshots there might be different offsets between wal index and raft entry index
-        let offset = first_entry.index - self.0.first_index();
-        <RaftEntry as prost_for_raft::Message>::decode(
-            self.0
-                .entry(id - offset)
-                .ok_or(raft::Error::Store(raft::StorageError::Unavailable))?
-                .as_ref(),
-        )
-        .map_err(consensus_manager::raft_error_other)
+
+        let wal_index = self
+            .index_offset()?
+            .try_raft_to_wal(raft_index)
+            .ok_or_else(|| raft::Error::Store(raft::StorageError::Compacted))?;
+
+        self.entry_by_wal_index(wal_index)
     }
 
     pub fn entries(
```

Comment:
Is `until` inclusive or exclusive? I thought it was inclusive before, and it looks like it's implemented to be exclusive now.

Either way, let's document it in a comment here.

Reply:
It's currently exclusive, and _I think_ I've made it that way, cause original implementation was exclusive. But I'll verify.

Reply:
> But I'll verify.

Thanks! :pray:

Reply:
[I think tt was exclusive.](https://github.com/qdrant/qdrant/pull/5217#discussion_r1796838494)